sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuScalarSubquery.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids
+
+import com.nvidia.spark.rapids.{GpuExec, GpuExpression}
+
+import org.apache.spark.TaskContext
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Expression, ExprId}
+import org.apache.spark.sql.execution.{BaseSubqueryExec, ExecSubqueryExpression, SparkPlan, UnaryExecNode}
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+case class GpuScalarSubquery(
+    plan: BaseSubqueryExec,
+    exprId: ExprId)
+  extends ExecSubqueryExpression with GpuExpression {
+
+  override def dataType: DataType = plan.schema.fields.head.dataType
+  override def children: Seq[Expression] = Nil
+  override def nullable: Boolean = true
+  override def withNewPlan(query: BaseSubqueryExec): GpuScalarSubquery = copy(plan = query)
+
+  override def semanticEquals(other: Expression): Boolean = other match {
+    case s: GpuScalarSubquery => plan.sameResult(s.plan)
+    case _ => false
+  }
+  private var result: Any = _
+
+  override def updateResult(): Unit = {
+    val subPlanRDD = plan.child.execute()
```

Comment:
Why are we doing this so differently from the Spark version?  It looks like we really wanted to replace the `BaseSubqueryExec`, and to work around replacing it we had to jump through a number of hoops making the code diverge from the original Spark code.  I really would prefer to just do it the same way that Spark did. I don't know 100% what types of `BaseSubqueryExec` can be passed into this, but if it ends up being a `SubqueryBroadcastExec` or a `ReusedSubqueryExec` this code will fail.

Reply:
Yes, I checked catalyst codes and found that `ReusedSubqueryExec`  may also be passed into `ScalarSubquery`. The intention of replacing original Spark code is to avoid `executeCollect` which seems inefficient. I realized this replacement seems unnecessary. So, I rewrote `updateResult` with the exact Spark code.